lib/backend/k8sv2/conversion.go

```diff
@@ -0,0 +1,431 @@
+		return false
+	} else if !c.isScheduled(pod) {
+		log.WithField("pod", pod.Name).Debug("Pod is not scheduled.")
+		return false
+	}
+	return true
+}
+
+func (c Converter) isScheduled(pod *kapiv1.Pod) bool {
+	return pod.Spec.NodeName != ""
+}
+
+func (c Converter) isHostNetworked(pod *kapiv1.Pod) bool {
+	return pod.Spec.HostNetwork
+}
+
+func (c Converter) hasIPAddress(pod *kapiv1.Pod) bool {
+	return pod.Status.PodIP != ""
+}
+
+// PodToWorkloadEndpoint converts a Pod to a WorkloadEndpoint.  It assumes the calling code
+// has verified that the provided Pod is valid to convert to a WorkloadEndpoint.
+func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error) {
+	// Pull out the profile and workload ID based on pod name and Namespace.
+	profile := fmt.Sprintf("k8s_ns.%s", pod.ObjectMeta.Namespace)
+	workload := fmt.Sprintf("%s.%s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
+
+	// We do, in some circumstances, want to parse Pods without an IP address.  For example,
+	// a DELETE update will not include an IP.
+	ipNets := []string{}
+	if c.hasIPAddress(pod) {
+		_, ipNet, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", pod.Status.PodIP))
+		if err != nil {
+			log.WithFields(log.Fields{"ip": pod.Status.PodIP, "pod": pod.Name}).WithError(err).Error("Failed to parse pod IP")
+			return nil, err
+		}
+		ipNets = append(ipNets, ipNet.String())
+	}
+
+	// Generate the interface name and MAC based on workload.  This must match
+	// the host-side veth configured by the CNI plugin.
+	interfaceName := VethNameForWorkload(workload)
+
+	// Build the labels map.
+	labels := map[string]string{}
+	if pod.ObjectMeta.Labels != nil {
+		labels = pod.ObjectMeta.Labels
+	}
+	labels["calico/k8s_ns"] = pod.ObjectMeta.Namespace
+
+	// TODO(doublek): Where do named ports go?
+	// Map any named ports through.
+	//var endpointPorts []model.EndpointPort
+	//for _, container := range pod.Spec.Containers {
+	//	for _, containerPort := range container.Ports {
+	//		if containerPort.Name != "" && containerPort.ContainerPort != 0 {
+	//			var modelProto numorstring.Protocol
+	//			switch containerPort.Protocol {
+	//			case kapiv1.ProtocolUDP:
+	//				modelProto = numorstring.ProtocolFromString("udp")
+	//			case kapiv1.ProtocolTCP, kapiv1.Protocol("") /* K8s default is TCP. */ :
+	//				modelProto = numorstring.ProtocolFromString("tcp")
+	//			default:
+	//				log.WithFields(log.Fields{
+	//					"protocol": containerPort.Protocol,
+	//					"pod":      pod,
+	//					"port":     containerPort,
+	//				}).Debug("Ignoring named port with unknown protocol")
+	//				continue
+	//			}
+
+	//			endpointPorts = append(endpointPorts, model.EndpointPort{
+	//				Name:     containerPort.Name,
+	//				Protocol: modelProto,
+	//				Port:     uint16(containerPort.ContainerPort),
+	//			})
+	//		}
+	//	}
+	//}
+
+	// Create the key / value pair to return.
+	// TODO(doublek): Which container becomes ContainerID in WEPSpec?
```

Comment:
Yeah, strictly speaking we can't get the container ID when running in KDD mode.

We probably need to leave that blank, or we need the CNI plugin to apply an annotation which stores it on the Pod.

Reply:
@caseydavenport You are correct. IIUC, it should be required for name construction only when using the etcd backend. Will remove the TODO.